Userland/Libraries/LibJS/Runtime/IteratorOperations.cpp

```diff
@@ -69,9 +70,38 @@ Object* iterator_next(Object& iterator, Value value)
 }
 
 // 7.4.6 IteratorClose ( iteratorRecord, completion ), https://tc39.es/ecma262/#sec-iteratorclose
-void iterator_close([[maybe_unused]] Object& iterator)
+void iterator_close(Object& iterator)
 {
-    TODO();
+    auto& vm = iterator.vm();
+    auto& global_object = iterator.global_object();
+
+    // Emulates `completion` behaviour
+    auto* exception = vm.exception();
+    vm.clear_exception();
+    auto unwind_until = vm.unwind_until();
+    auto unwind_until_label = vm.unwind_until_label();
+    vm.stop_unwind();
+    ArmedScopeGuard restore_exception = [&]() {
+        if (exception)
+            vm.set_exception(*exception);
+        if (unwind_until != ScopeType::None)
+            vm.unwind(unwind_until, unwind_until_label);
+    };
+
+    auto return_method = get_method(global_object, &iterator, vm.names.return_);
+    if (!return_method)
+        return;
+
+    auto result = vm.call(*return_method, &iterator);
+    if (vm.exception())
+        return;
+    if (exception)
+        return;
+    restore_exception.disarm();
```

Comment:
This seems incorrect:

```
6. If completion.[[Type]] is throw, return Completion(completion).
7. If innerResult.[[Type]] is throw, return Completion(innerResult).
```

I would expect something like:

```suggestion
    if (exception)
        return;
    restore_exception.disarm();
    if (vm.exception())
        return;
```

as `exception` is for `completion` and `vm.exception()` for `innerResult`, no?

Reply:
While i will change it to that order to more clearly match the spec, in this scenario theyre actually equivalent, since the restore exception guard was not disarmed yet. (so in both cases it returns the completion if it is a throw :)

Reply:
Exactly, but in the `vm.exception()` case it shouldn't restore the exception - `return Completion(innerResult)`. That's an exception from `Set innerResult to Call(return, iterator).`

Reply:
right, but the `if (exception)` check in the restore_exception wouldnt overwrite the inner result exception if the completion is not a throw. I think this scope guard implementation is a bit opaque, so ill try to rewrite it a bit more clearly.

Reply:
ohh, I totally missed that check. `restore_exception` sounds very unconditional :sweat_smile:

Reply:
rewritten :)